File: source/equipment/echelon.py
#
# MPowerTCX: Share Schwinn A.C. indoor cycle data with Strava, GoldenCheetah and other apps
# Copyright (C) 2017 James Roth
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import re

from . import bikes

#
# The version numbers represent the order in which I encountered them, not official firmware revisions
#
class EchelonV1(bikes.Bike):
    def load(self, peek, reader):
        if peek == ['Stage_Totals']:
            self._load_header(reader)
            return True
        elif peek == ['Stage_Workout (min)', 'Distance(km)', 'Speed(km/h)', 'Watts ', 'HR ', 'RPM ']:
            self._load_data(reader)
            return True
            
        return False

    def name(self):
        return "Echelon Variant 1"
        
    def _load_header(self, reader):
        header = {}

        for row in reader:
            if len(row):
                header[row[0]] = row[1]
            else:
                break

        self.ride.header.setSummary(
            time=float(header["Total Time"]) * 60.0,
            distance=float(header["Total_distance:"]) * 1000.0,
            average_power=header["Watts Avg"],
            max_power=header["Watts Max"],
            average_rpm=header["RPM Avg"],
            max_rpm=header["RPM Max"],
            average_hr=header["HR Avg"],
            max_hr=header["HR Max"],
            calories=header["KCal"]
        )

    def _load_data(self, reader):
        last_time = 0
        
        for row in reader:
            if len(row) == 0:
                break
            elif len(row) == 6:
                last_time = float(row[0]) * 60
                self.ride.addSample(
                    power=row[3],
                    rpm=row[5],
                    hr=row[4],
                    distance=float(row[1]) * 1000.0
                )
            else:
                self.skip(row)

        if self.ride.header.time == 0:
            print ("v1 header missing")
            self.ride.inferHeader(last_time)
    
    
class EchelonV2(bikes.Bike):
    def load(self, peek, reader):
        if peek[0:2] == ['RIDE SUMMARY', '']:
            self._load_header(reader)
            return True
        elif peek[0:2] == ['RIDE DATA', '']:
            self._load_data(reader)
            return True
        elif re.match('^STAGE_[0-9]+_SUMMARY$', peek[0]):
            self._skip_section(reader)
            return True
            
        return False

    def name(self):
        return "Echelon Variant 2"
        
    def _skip_section(self, reader):
        for row in reader:
            # Return on "" or ",,,"
            if len(row) == 0 or ''.join(row) == '':
                break
    
    def _load_header(self, reader):
        header = {}

        for row in reader:
#            print ('ROW %r' % row)
            
            # Return on "" or ",,,"
            if len(row) and ''.join(row) != '':
                header[row[0]] = row[1]
            else:
                break

#        print ('HEADER %r' % header)
        
        self.ride.header.setSummary(
            time=float(header["Total Time"]) * 60.0,
            distance=float(header["Total Distance"]) * 1609.34,
            average_power=header["AVG Power"],
            max_power=header["MAX Power"],
            average_rpm=header["AVG RPM"],
            max_rpm=header["MAX RPM"],
            average_hr=header["AVG HR"],
            max_hr=header["MAX HR"],
            calories=header["CAL"]
        )

    def _load_data(self, reader):
        keys = next(reader)

        for row in reader:
            if len(row):
                data = dict(zip(keys, row))
                self.ride.addSample(
                    power=data["Power"],
                    rpm=data["RPM"],
                    hr=data["HR"],
                    distance=float(data["DISTANCE"]) * 1609.34
                )
            else:
                break

                
class EchelonV3(bikes.Bike):
    """ 
    So called v3 is a messed up version of v1. I suspect it's an earlier firmware. 
    """
    def load(self, peek, reader):
        if peek == ['Stage_Workout (min)', 'Distance(mile)', 'Speed (mph)', 'Watts ', 'HR ', 'RPM ']:
            self._load(reader)
            return True
            
        return False

    def name(self):
        return "Echelon Variant 3"
        
    def _load(self, reader):
        for row in reader:
            if len(row) == 6:
                self.ride.addSample(
                    power=row[3],
                    rpm=row[5],
                    hr=row[4],
                    distance=float(row[1]) * 1609.34
                )
            elif row == ['Stage_Totals']:
                self._load_header(reader)
            else:
                self.skip(row)
        
    def _load_header(self, reader):
        header = {}

        for row in reader:
            if len(row):
                header[row[0]] = row[1]
            else:
                break

        self.ride.header.setSummary(
            time=float(header["Total Time"]) * 60.0,
            distance=float(header["Total_distance:"]) * 1609.34,
            average_power=header["Watts Avg"],
            max_power=header["Watts Max"],
            average_rpm=header["RPM Avg"],
            max_rpm=header["RPM Max"],
            average_hr=header["HR Avg"],
            max_hr=header["HR Max"],
            calories=header["KCal"]
        )
     

File: crates/mpowertcx-core/src/equipment/echelon.rs
use crate::equipment::{BikeParser, CsvRows};
use crate::ride::{python_float, Ride};

pub struct EchelonV1;

impl EchelonV1 {
    pub fn new() -> Self {
        Self
    }

    fn load_header(rows: &mut CsvRows, ride: &mut Ride) {
        let mut header = std::collections::HashMap::new();
        while let Some(row) = rows.next() {
            if row.is_empty() {
                break;
            }
            if row.len() >= 2 {
                header.insert(row[0].clone(), row[1].clone());
            }
        }

        let time = header.get("Total Time").and_then(|v| v.parse::<f64>().ok()).unwrap_or(0.0) * 60.0;
        let distance = header.get("Total_distance:").and_then(|v| v.parse::<f64>().ok()).unwrap_or(0.0) * 1000.0;

        ride.header.set_summary(
            time,
            python_float(time),
            distance,
            header.get("Watts Avg").cloned().unwrap_or_default(),
            header.get("Watts Max").cloned().unwrap_or_default(),
            header.get("RPM Avg").cloned().unwrap_or_default(),
            header.get("RPM Max").cloned().unwrap_or_default(),
            header.get("HR Avg").cloned().unwrap_or_default(),
            header.get("HR Max").cloned().unwrap_or_default(),
            header.get("KCal").cloned().unwrap_or_default(),
        );
    }

    fn load_data(rows: &mut CsvRows, ride: &mut Ride) {
        let mut last_time = 0.0f64;
        while let Some(row) = rows.next() {
            if row.is_empty() {
                break;
            }
            if row.len() == 6 {
                last_time = row[0].parse::<f64>().unwrap_or(0.0) * 60.0;
                let distance = row[1].parse::<f64>().unwrap_or(0.0) * 1000.0;
                ride.add_sample(
                    &row[3],
                    &row[5],
                    &row[4],
                    distance,
                );
            }
        }

        if ride.header.time == 0.0 {
            ride.infer_header(last_time, python_float(last_time));
        }
    }
}

impl BikeParser for EchelonV1 {
    fn try_load(&mut self, peek: &[String], rows: &mut CsvRows, ride: &mut Ride) -> Result<bool, String> {
        if peek.len() == 1 && peek[0] == "Stage_Totals" {
            Self::load_header(rows, ride);
            return Ok(true);
        }
        let expected = ["Stage_Workout (min)", "Distance(km)", "Speed(km/h)", "Watts ", "HR ", "RPM "];
        if peek.len() == 6 && peek.iter().zip(expected.iter()).all(|(a, b)| a == b) {
            Self::load_data(rows, ride);
            return Ok(true);
        }
        Ok(false)
    }

    fn name(&self) -> &str {
        "Echelon Variant 1"
    }

    fn slug(&self) -> &str {
        "echelon-v1"
    }

    fn incline_is_simulated(&self) -> bool {
        true
    }
}

pub struct EchelonV2;

impl EchelonV2 {
    pub fn new() -> Self {
        Self
    }

    fn is_stage_summary(peek: &[String]) -> bool {
        if peek.is_empty() {
            return false;
        }
        let s = &peek[0];
        s.starts_with("STAGE_") && s.ends_with("_SUMMARY") && {
            let mid = &s[6..s.len() - 8];
            !mid.is_empty() && mid.chars().all(|c| c.is_ascii_digit())
        }
    }

    fn skip_section(rows: &mut CsvRows) {
        while let Some(row) = rows.next() {
            if row.is_empty() || row.iter().all(|s| s.is_empty()) {
                break;
            }
        }
    }

    fn load_header(rows: &mut CsvRows, ride: &mut Ride) {
        let mut header = std::collections::HashMap::new();
        while let Some(row) = rows.next() {
            if row.is_empty() || row.iter().all(|s| s.is_empty()) {
                break;
            }
            if row.len() >= 2 {
                header.insert(row[0].clone(), row[1].clone());
            }
        }

        let time = header.get("Total Time").and_then(|v| v.parse::<f64>().ok()).unwrap_or(0.0) * 60.0;
        let distance = header.get("Total Distance").and_then(|v| v.parse::<f64>().ok()).unwrap_or(0.0) * 1609.34;

        ride.header.set_summary(
            time,
            python_float(time),
            distance,
            header.get("AVG Power").cloned().unwrap_or_default(),
            header.get("MAX Power").cloned().unwrap_or_default(),
            header.get("AVG RPM").cloned().unwrap_or_default(),
            header.get("MAX RPM").cloned().unwrap_or_default(),
            header.get("AVG HR").cloned().unwrap_or_default(),
            header.get("MAX HR").cloned().unwrap_or_default(),
            header.get("CAL").cloned().unwrap_or_default(),
        );
    }

    fn load_data(rows: &mut CsvRows, ride: &mut Ride) {
        let keys = rows.next().cloned().unwrap_or_default();
        while let Some(row) = rows.next() {
            if row.is_empty() {
                break;
            }
            let data: std::collections::HashMap<&str, &str> = keys
                .iter()
                .zip(row.iter())
                .map(|(k, v)| (k.as_str(), v.as_str()))
                .collect();
            let distance = data.get("DISTANCE").and_then(|v| v.parse::<f64>().ok()).unwrap_or(0.0) * 1609.34;
            ride.add_sample(
                data.get("Power").copied().unwrap_or("0"),
                data.get("RPM").copied().unwrap_or("0"),
                data.get("HR").copied().unwrap_or("0"),
                distance,
            );
        }
    }
}

impl BikeParser for EchelonV2 {
    fn try_load(&mut self, peek: &[String], rows: &mut CsvRows, ride: &mut Ride) -> Result<bool, String> {
        if peek.len() >= 2 && peek[0] == "RIDE SUMMARY" && peek[1] == "" {
            Self::load_header(rows, ride);
            return Ok(true);
        }
        if peek.len() >= 2 && peek[0] == "RIDE DATA" && peek[1] == "" {
            Self::load_data(rows, ride);
            return Ok(true);
        }
        if Self::is_stage_summary(peek) {
            Self::skip_section(rows);
            return Ok(true);
        }
        Ok(false)
    }

    fn name(&self) -> &str {
        "Echelon Variant 2"
    }

    fn slug(&self) -> &str {
        "echelon-v2"
    }

    fn incline_is_simulated(&self) -> bool {
        true
    }
}

pub struct EchelonV3;

impl EchelonV3 {
    pub fn new() -> Self {
        Self
    }

    fn load_header(rows: &mut CsvRows, ride: &mut Ride) {
        let mut header = std::collections::HashMap::new();
        while let Some(row) = rows.next() {
            if row.is_empty() {
                break;
            }
            if row.len() >= 2 {
                header.insert(row[0].clone(), row[1].clone());
            }
        }

        let time = header.get("Total Time").and_then(|v| v.parse::<f64>().ok()).unwrap_or(0.0) * 60.0;
        let distance = header.get("Total_distance:").and_then(|v| v.parse::<f64>().ok()).unwrap_or(0.0) * 1609.34;

        ride.header.set_summary(
            time,
            python_float(time),
            distance,
            header.get("Watts Avg").cloned().unwrap_or_default(),
            header.get("Watts Max").cloned().unwrap_or_default(),
            header.get("RPM Avg").cloned().unwrap_or_default(),
            header.get("RPM Max").cloned().unwrap_or_default(),
            header.get("HR Avg").cloned().unwrap_or_default(),
            header.get("HR Max").cloned().unwrap_or_default(),
            header.get("KCal").cloned().unwrap_or_default(),
        );
    }

    fn load(rows: &mut CsvRows, ride: &mut Ride) {
        while let Some(row) = rows.next() {
            if row.len() == 6 {
                let distance = row[1].parse::<f64>().unwrap_or(0.0) * 1609.34;
                ride.add_sample(&row[3], &row[5], &row[4], distance);
            } else if row.len() == 1 && row[0] == "Stage_Totals" {
                Self::load_header(rows, ride);
            }
        }
    }
}

impl BikeParser for EchelonV3 {
    fn try_load(&mut self, peek: &[String], rows: &mut CsvRows, ride: &mut Ride) -> Result<bool, String> {
        let expected = ["Stage_Workout (min)", "Distance(mile)", "Speed (mph)", "Watts ", "HR ", "RPM "];
        if peek.len() == 6 && peek.iter().zip(expected.iter()).all(|(a, b)| a == b) {
            Self::load(rows, ride);
            return Ok(true);
        }
        Ok(false)
    }

    fn name(&self) -> &str {
        "Echelon Variant 3"
    }

    fn slug(&self) -> &str {
        "echelon-v3"
    }

    fn incline_is_simulated(&self) -> bool {
        true
    }
}
